### PR TITLE
Add typed client options

### LINKED ISSUE #50: Type Hints for Supabase-py
It would be nice if we could get type hints on the library so that we can allow users to sanity check their code.

Thank you @dreinon for proposing this!

### PR DESCRIPTION
**Pull request by joeriddles**
Closes #50 

This PR: 
- Adds more type hints throughout.
- Adds typed supabase `Client` options in the form of `ClientOptions`.
- Adds a `test.ps1` file that is equivalent to `test.sh`, to more easily run tests on Windows.

### CODE REVIEW

**Comment by joeriddles** [comment]
I ran pre-commit on my end and everything is looking good. I did have to make a commit to fix all the line-endings. Setting `core.autocrlf = true` should fix this in the future.
```
> pre-commit run --all-files
pre-commit run --all-files
Trim Trailing Whitespace.................................................Passed
Check for added large files..............................................Passed
Mixed line ending........................................................Passed
isort....................................................................Passed
autoflake................................................................Passed
black....................................................................Passed
```

**Comment by dreinon** [comment]
@joeriddles thanks for the contribution! Tests are failing, could you take a look whenever you have time, please?

Thanks!

**Comment by joeriddles** [comment]
@dreinon fixed the tests! Looks like they're passing on my machine. I'm not exactly sure what `xfailed` and `xpassed` mean, but they're not red.
```
collected 57 items

tests\test_client.py xxxxxxxxxxxxxxxxXXXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx...s                                                                                     [ 92%]
tests\test_client_options.py ..                                                                                                                                [ 96%] 
tests\test_dummy.py ..                                                                                                                                         [100%]

=== 7 passed, 1 skipped, 46 xfailed, 3 xpassed in 4.67s ===
```

**Comment by dreinon** [comment]
@leynier @anand2312 can you review this too please?

**Review by dreinon — APPROVED** [review]

**Comment by dreinon** [comment]
> @dreinon fixed the tests! Looks like they're passing on my machine. I'm not exactly sure what `xfailed` and `xpassed` mean, but they're not red.
> ```
> collected 57 items
> 
> tests\test_client.py xxxxxxxxxxxxxxxxXXXxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx...s                                                                                     [ 92%]
> tests\test_client_options.py ..                                                                                                                                [ 96%] 
> tests\test_dummy.py ..                                                                                                                                         [100%]
> 
> === 7 passed, 1 skipped, 46 xfailed, 3 xpassed in 4.67s ===
> ```

xfails are tests that are expected to fail, but xpasses are tests that are expected to fail but passed (which shouldn't be the case). Could you look into these cases too please @joeriddles?

**Comment by dreinon** [comment]
Well, actually, these xpasses are happening in a file this PR does not touch, so no worries about that.

**Comment by dreinon** [comment]
What I would actually like you to check is the two refactors sourcery proposes in PR #97, last commit. These cases are when variables are inmediately returned, sourcery asks you to return its value directly.

File: supabase/lib/client_options.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    """Whether to persist a logged in session to storage."""
+    persist_session: bool = True
+
+    """Detect a session from the URL. Used for OAuth login callbacks."""
+    detect_session_in_url: bool = True
+
+    """A storage provider. Used to store the logged in session."""
+    local_storage: Dict[str, Any] = dataclasses.field(default_factory=lambda: {})
+
+    """Options passed to the realtime-js instance"""
```
[the remark this answers is not in the document]
Comment (joeriddles) [reply supabase/lib/client_options.py:32]: > Realtime-py?

Not sure I understand this comment. Is `client_options` missing some realtime_py values?

**Comment by joeriddles** [comment]
> What I would actually like you to check is the two refactors sourcery proposes in PR #97, last commit. These cases are when variables are inmediately returned, sourcery asks you to return its value directly.

Sure, made the changes, though one of sourcery's suggested changes was for part of a function I didn't touch. Regardless, made the small tweaks.

File: supabase/lib/client_options.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anand2312) [reply supabase/lib/client_options.py:32]: The existing comment says `realtime-js`, I'm guessing dreinon is asking you to change that to realtime-py

File: supabase/client.py [hunk 2]
```diff
@@ -100,7 +94,7 @@ def from_(self, table_name: str):
         )
         return query_builder.from_(table_name)
 
-    def rpc(self, fn, params):
+    def rpc(self, fn, params) -> Coroutine[Any, Any, Response]:
```
Comment (anand2312) [inline supabase/client.py:97]:
```suggestion
    def rpc(self, fn: str, params: Dict[Any, Any]) -> Coroutine[Any, Any, Response]:
```
Missing parameter annotations

File: supabase/lib/storage_client.py [hunk 3]
```diff
@@ -14,8 +16,8 @@ class SupabaseStorageClient(StorageBucketAPI):
     >>> list_files = storage_file.list("something")
     """
 
-    def __init__(self, url, headers):
+    def __init__(self, url: str, headers: Dict[str, str]):
         super().__init__(url, headers)
 
-    def StorageFileAPI(self, id_):
+    def StorageFileAPI(self, id_: str):
```
Comment (anand2312) [inline supabase/lib/storage_client.py:22]:
```suggestion
    def StorageFileAPI(self, id_: str) -> StorageFileAPI:
```
Missing return type annotation

**Review by J0 — APPROVED** [review]

**Review by dreinon — APPROVED** [review]

**Comment by joeriddles** [comment]
@anand2312 thanks for adding those type hints

**Comment by dreinon** [comment]
Perfect, so let's wait for @leynier 's review and then we'll ve ready to merge :)

**Comment by joeriddles** [comment]
I also added `py.typed` to implement [PEP561](https://www.python.org/dev/peps/pep-0561/), "Distributing and Packaging Type Information".

**Review by leynier — APPROVED** [review]

**Comment by leynier** [comment]
LGTM! Thanks for the contribution @joeriddles

**Comment by dreinon** [comment]
Tests are failing due to realtime-py dependency

**Pull request merged**